docs/plots/backends.rst

```diff
@@ -0,0 +1,362 @@
+************************
+
+.. warning::
+   The plotting backend API is currently re-designed and details might change in the near future.
+
+Plotting in Sherpa is done through *plotting backends*. A *backend* is an
+external package that performs the actual plotting function such as :term:`matplotlib`, i.e. it creates a
+canvas and puts color on paper or on the screen. The Sherpa plotting classes are
+written in a backend-independent way; they work with any plotting backend that
+is supported by Sherpa. We also use the term *backend* to refer to Sherpa
+classes that connect the Sherpa plot objects with the plotting backend by
+translating Sherpa options into backend-specific commands. The details of that
+are explained below, but, in short, a backend class would translate the sherpa
+backend-independent "plot a line in red from here to there" into backend
+specific ``plt.plot(x,y, linecolor='r')`` (if the backend is :term:`matplotlib`).
+
+Instead of utilizing Sherpa classes and commands to plot data, users can
+alternatively just access the data in the Sherpa data objects (e.g. the x and y
+values of a dataset) and perform plotting operations directly with any plotting
+backend that is available to them. This method may be less convenient, but it
+works for ploting backends not (yet) supported by Sherpa.
+
+Which backend is used?
+======================
+
+When the `sherpa.plot` module is first imported, Sherpa tries to import the
+backends installed with Sherpa in the order listed in the
+``options.plot_pkg`` setting from the ``sherpa.rc`` startup file.
+The first module that imports successfully is set as the active
+backend. The following command prints the name and the location
+on disk of that module::
+
+   >>> from sherpa import plot
+   >>> print(plot.backend)
+
+Change the backend
+------------------
+
+.. warning::
+   When the plotting backend is changed, new plotting objects must be
+   initialized because objects like e.g. `sherpa.plot.DataPlot` keep a
+   copy of all default parameters of the backend that was active when they
+   are initialized. A different backend might not understand all those
+   parameters.
+
+   In particular, that chnging the backend might not work for plot functions
+   in Sherpa's UI layer such as `sherpa.astro.plot.plot_data` because those
+   work with objects underneath that are not updated when the plotting backend
+   is changed.
+
+After the initial import, the backend can be changed by loading one of
+the plotting backends shipped with sherpa (or any other module that
+provides the same interface)::
+
+  >>> import sherpa.plot.pylab_backend
+  >>> plot.backend = sherpa.plot.pylab_backend
+
+Sherpa also provides a decorator to change the backend for just one plot::
+
+  >>> from sherpa.plot.backends import TemporaryPlottingBackend
+  >>> from sherpa.plot.pylab_backend import PylabBackend
+  >>> with TemporarypPlottingBackend(PylabBackend()):
+  ...     plotting code here...
+
+
```

Comment:
Yay, I was able to get this working (I don't have a `plot_pkg` line in `~/.sherpa-standalone.rc` here so the default is the dummy plot backend):

```
In [1]: from sherpa.astro.ui import *
WARNING: imaging routines will not be available,
failed to import sherpa.image.ds9_backend due to
'RuntimeErr: DS9Win unusable: Could not find ds9 on your PATH'
WARNING: failed to import sherpa.astro.xspec; XSPEC models will not be available

In [2]: load_arrays(1, [1, 2, 3], [10, 30, 20])

In [3]: plot_data()
WARNING: <class 'sherpa.plot.backends.BasicBackend'> does not implement line/symbol plotting. No plot will be produced.

In [4]: from sherpa.plot import TemporaryPlottingBackend

In [5]: from sherpa.plot.pylab_backend import PylabBackend

In [6]: with TemporaryPlottingBackend(PylabBackend()):
   ...:     plot_data()
   ...:


**AT THIS POINT THE MATPLOTLIB WINDOW APPEARED***

In [7]: plot_data()
WARNING: <class 'sherpa.plot.backends.BasicBackend'> does not implement line/symbol plotting. No plot will be produced.

```

Reply:
Interesting. I now get the following

```
% cat delme.rc 
[options]
plot_pkg : none
% SHERPARC=delme.rc ipython --matplotlib
...
In [1]: from sherpa.astro.ui import *
WARNING: Plotting backend 'none' not found or dependencies missing. Trying next option.
WARNING: Tried the following backends listed in delme.rc: 
['none']
None of these imported correctly, so using the 'BasicBackend'.
List of backends that have loaded and would be available: ['BaseBackend', 'BasicBackend', 'IndepOnlyBackend', 'pylab', 'PylabErrorArea']
WARNING: Cannot import usable I/O backend.
    If you are using CIAO, this is most likely an error and you should contact the CIAO helpdesk.
    If you are using Standalone Sherpa, please install astropy.

In [2]: load_arrays(1, [1, 2, 3], [10, 30, 20])

In [3]: plot_data()
WARNING: <class 'sherpa.plot.backends.BasicBackend'> does not implement line/symbol plotting. No plot will be produced.

In [4]: dp = get_data_plot()

In [5]: dp.x, dp.y
Out[5]: (array([1, 2, 3]), array([10, 30, 20]))

In [6]: from sherpa.plot import TemporaryPlottingBackend

In [7]: from sherpa.plot.pylab_backend import PylabBackend

In [8]: with TemporaryPlottingBackend(PylabBackend()):
    ...:     plot_data()
    ...: 
WARNING: Plotting backend not changed because <sherpa.plot.pylab_backend.PylabBackend object at 0x7fae10cdd890> is not a known plotting backend.
List of available plotting backends: ['BaseBackend', 'BasicBackend', 'IndepOnlyBackend', 'pylab', 'PylabErrorArea']
WARNING: <class 'sherpa.plot.backends.BasicBackend'> does not implement line/symbol plotting. No plot will be produced.

In [9]: with TemporaryPlottingBackend("pylab"):
    ...:     plot_data()
    ...: 
``` 


So, we can no longer pass a class to the `TemporaryPlottingBackend` cnotext manager. I wonder if we would ever want this?